packages/components/src/components/Calculator/CalculatorResult.tsx

```diff
@@ -0,0 +1,69 @@
+import { FC, useEffect, useMemo, useState } from "react";
+
+import { Env, SqValue } from "@quri/squiggle-lang";
+
+import { PlaygroundSettings } from "../PlaygroundSettings.js";
+import { ValueResultViewer } from "./ValueResultViewer.js";
+import {
+  InputValues,
+  SqCalculatorValueWithContext,
+  SqValueResult,
+} from "./types.js";
+import { useSavedCalculatorState } from "./useSavedCalculatorState.js";
+
+type Props = {
+  valueWithContext: SqCalculatorValueWithContext;
+  inputValues: InputValues;
+  environment: Env;
+  settings: PlaygroundSettings;
+};
+
+export const CalculatorResult: FC<Props> = ({
+  valueWithContext,
+  inputValues,
+  environment,
+  settings,
+}) => {
+  const [cachedState, updateCachedState] =
+    useSavedCalculatorState(valueWithContext);
+
+  const calculator = useMemo(() => valueWithContext.value, [valueWithContext]);
+
+  const [fnValue, setFnValue] = useState<SqValueResult | undefined>(
+    () => cachedState?.fnValue
+  );
+
+  // Whenever `inputValues` change, we recalculate `fnValue`.
+  useEffect(() => {
+    const parameters: SqValue[] = [];
+
+    // Unpack all input values.
+    for (const input of calculator.inputs) {
+      const inputValue = inputValues[input.name];
+      if (!inputValue || !inputValue.ok) {
+        // One of inputs is incorrect.
+        setFnValue(undefined);
+        return;
+      }
+      parameters.push(inputValue.value);
+    }
+
+    const finalResult = calculator.run(parameters, environment);
+    setFnValue(finalResult);
```

Comment:
I feel kind of uncomfortable with the state/calculations for calculating/setting inputValues being so far removed from that for the fnValue, but perhaps that can be improved later.

Reply:
I don't see why that's a problem? `<Calculator> = <Header> + <Inputs> + <Result>` seems like a natural decomposition to me, with inputs feeding to result to calculate `fnValue`.

One awkward thing here is that we now have to backup `fnValue` and `inputValues` separately, with some minor overhead and discrepancy risk because we call `useSavedCalculatorState` twice (and so do `sameCalculatorCacheExists` check twice),  but in practice it works well.

Also, calculating `fnValue` in a separate component is useful to avoid re-renders, as I mentioned in another comment.